File: phases/09-computer-networks/13-http-1-1-http-2-wire-format-and-multiplexing/code/main.py
"""HTTP/1.1 and HTTP/2 protocol utilities."""

from dataclasses import dataclass, field
from typing import Optional


# --- HTTP/1.1 ---

@dataclass
class HttpRequest:
    method: str
    path: str
    version: str
    headers: dict[str, str] = field(default_factory=dict)
    body: bytes = b""


@dataclass
class HttpResponse:
    status: int
    reason: str
    headers: dict[str, str] = field(default_factory=dict)
    body: bytes = b""


def parse_http_request(data: bytes) -> HttpRequest:
    """Parse raw bytes into an HTTP/1.1 request."""
    text = data.decode("utf-8", errors="replace")
    parts = text.split("\r\n\r\n", 1)
    header_section = parts[0]
    body = parts[1].encode() if len(parts) > 1 else b""

    lines = header_section.split("\r\n")
    request_line = lines[0].split(" ", 2)
    if len(request_line) < 3:
        raise ValueError(f"Invalid request line: {lines[0]}")

    method, path, version = request_line
    headers = {}
    for line in lines[1:]:
        if ":" in line:
            key, value = line.split(":", 1)
            headers[key.strip().lower()] = value.strip()

    return HttpRequest(method=method, path=path, version=version, headers=headers, body=body)


def build_http_response(response: HttpResponse) -> bytes:
    """Serialize an HTTP response to bytes."""
    lines = [f"HTTP/1.1 {response.status} {response.reason}"]
    for key, value in response.headers.items():
        lines.append(f"{key}: {value}")
    header_block = "\r\n".join(lines) + "\r\n\r\n"
    return header_block.encode() + response.body


def format_chunked(body: bytes, chunk_size: int = 16) -> bytes:
    """Encode a body using chunked transfer encoding."""
    result = []
    for i in range(0, len(body), chunk_size):
        chunk = body[i : i + chunk_size]
        result.append(f"{len(chunk):X}\r\n".encode() + chunk + b"\r\n")
    result.append(b"0\r\n\r\n")
    return b"".join(result)


# --- HTTP/2 ---

H2_FRAME_TYPES = {
    0x0: "DATA",
    0x1: "HEADERS",
    0x2: "PRIORITY",
    0x3: "RST_STREAM",
    0x4: "SETTINGS",
    0x5: "PUSH_PROMISE",
    0x6: "PING",
    0x7: "GOAWAY",
    0x8: "WINDOW_UPDATE",
    0x9: "CONTINUATION",
}

H2_PREFACE = b"PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n"


@dataclass
class H2Frame:
    length: int
    frame_type: int
    flags: int
    stream_id: int
    payload: bytes

    @property
    def type_name(self) -> str:
        return H2_FRAME_TYPES.get(self.frame_type, "UNKNOWN")

    @property
    def has_ack(self) -> bool:
        return bool(self.flags & 0x01)

    @property
    def is_end_stream(self) -> bool:
        return bool(self.flags & 0x01)

    @property
    def is_end_headers(self) -> bool:
        return bool(self.flags & 0x04)


def parse_h2_frame(data: bytes) -> Optional[H2Frame]:
    """Parse an HTTP/2 frame from raw bytes. Returns None if insufficient data."""
    if len(data) < 9:
        return None

    length = int.from_bytes(data[0:3], "big")
    frame_type = data[3]
    flags = data[4]
    stream_id = int.from_bytes(data[5:9], "big") & 0x7FFFFFFF

    if len(data) < 9 + length:
        return None

    payload = data[9 : 9 + length]
    return H2Frame(length=length, frame_type=frame_type, flags=flags, stream_id=stream_id, payload=payload)


def build_h2_frame(frame_type: int, flags: int, stream_id: int, payload: bytes) -> bytes:
    """Build an HTTP/2 frame as bytes."""
    length = len(payload)
    header = bytearray(9)
    header[0:3] = length.to_bytes(3, "big")
    header[3] = frame_type
    header[4] = flags
    header[5:9] = (stream_id & 0x7FFFFFFF).to_bytes(4, "big")
    return bytes(header) + payload


def build_settings_frame(settings: dict[int, int] | None = None) -> bytes:
    """Build a SETTINGS frame."""
    defaults = {3: 100, 4: 65535, 5: 16384}
    if settings:
        defaults.update(settings)

    payload = b""
    for setting_id, value in defaults.items():
        payload += setting_id.to_bytes(2, "big") + value.to_bytes(4, "big")

    return build_h2_frame(0x04, 0x00, 0, payload)


def parse_settings_payload(payload: bytes) -> dict[int, int]:
    """Parse SETTINGS frame payload into a dict."""
    settings = {}
    for i in range(0, len(payload), 6):
        if i + 6 <= len(payload):
            setting_id = int.from_bytes(payload[i : i + 2], "big")
            value = int.from_bytes(payload[i + 2 : i + 6], "big")
            settings[setting_id] = value
    return settings


# --- HPACK Simulation ---

STATIC_TABLE = [
    (":authority", ""),
    (":method", "GET"),
    (":method", "POST"),
    (":path", "/"),
    (":path", "/index.html"),
    (":scheme", "http"),
    (":scheme", "https"),
    (":status", "200"),
    (":status", "204"),
    (":status", "206"),
    (":status", "304"),
    (":status", "400"),
    (":status", "404"),
    (":status", "500"),
    ("accept-charset", ""),
    ("accept-encoding", "gzip, deflate"),
    ("accept-language", ""),
    ("accept-ranges", ""),
    ("accept", ""),
    ("access-control-allow-origin", ""),
    ("age", ""),
    ("allow", ""),
    ("authorization", ""),
    ("cache-control", ""),
    ("content-disposition", ""),
    ("content-encoding", ""),
    ("content-language", ""),
    ("content-length", ""),
    ("content-location", ""),
    ("content-range", ""),
    ("content-type", ""),
    ("cookie", ""),
    ("date", ""),
    ("etag", ""),
    ("expect", ""),
    ("expires", ""),
    ("from", ""),
    ("host", ""),
    ("if-match", ""),
    ("if-modified-since", ""),
    ("if-none-match", ""),
    ("if-range", ""),
    ("if-unmodified-since", ""),
    ("last-modified", ""),
    ("link", ""),
    ("location", ""),
    ("max-forwards", ""),
    ("proxy-authenticate", ""),
    ("proxy-authorization", ""),
    ("range", ""),
    ("referer", ""),
    ("refresh", ""),
    ("retry-after", ""),
    ("server", ""),
    ("set-cookie", ""),
    ("strict-transport-security", ""),
    ("transfer-encoding", ""),
    ("user-agent", ""),
    ("vary", ""),
    ("via", ""),
    ("www-authenticate", ""),
]


def hpack_lookup(name: str, value: str) -> Optional[int]:
    """Look up a header in the HPACK static table. Returns index or None."""
    for i, (n, v) in enumerate(STATIC_TABLE, start=1):
        if n.lower() == name.lower() and (v == "" or v.lower() == value.lower()):
            return i
    return None


def hpack_compress(headers: dict[str, str]) -> list[tuple[str, str, Optional[int]]]:
    """Compress headers using static table lookup."""
    result = []
    for name, value in headers.items():
        index = hpack_lookup(name, value)
        result.append((name, value, index))
    return result


# --- Main Demo ---

def main():
    print("=== HTTP/1.1 and HTTP/2 Protocol Utilities ===\n")

    # HTTP/1.1 demo
    raw = b"POST /submit HTTP/1.1\r\nHost: example.com\r\nContent-Type: application/json\r\nContent-Length: 18\r\n\r\n{\"action\":\"create\"}"
    req = parse_http_request(raw)
    print(f"HTTP/1.1 Request: {req.method} {req.path} {req.version}")
    print(f"  Headers: {dict(req.headers)}")
    print(f"  Body: {req.body.decode()}")

    resp = HttpResponse(
        status=201,
        reason="Created",
        headers={"content-type": "application/json", "content-length": "15"},
        body=b'{"id": "abc123"}',
    )
    resp_bytes = build_http_response(resp)
    print(f"\nHTTP/1.1 Response:\n{resp_bytes.decode()}")

    # Chunked encoding demo
    chunked = format_chunked(b"Hello, chunked world!", chunk_size=8)
    print(f"Chunked encoding:\n{chunked.decode()}")

    # HTTP/2 frame demo
    settings = build_settings_frame()
    frame = parse_h2_frame(settings)
    if frame:
        print(f"HTTP/2 Frame: type={frame.type_name}, length={frame.length}, flags=0x{frame.flags:02X}, stream={frame.stream_id}")
        parsed_settings = parse_settings_payload(frame.payload)
        print(f"  Settings: {parsed_settings}")

    ping_payload = bytes.fromhex("DEADBEEF01020304")
    ping_frame = build_h2_frame(0x06, 0x00, 0, ping_payload)
    parsed = parse_h2_frame(ping_frame)
    if parsed:
        print(f"\nPING frame: payload={parsed.payload.hex()}")

    # HPACK demo
    print("\n--- HPACK Compression ---")
    headers = {":method": "GET", ":path": "/index.html", ":scheme": "https", "user-agent": "demo/1.0"}
    compressed = hpack_compress(headers)
    for name, value, index in compressed:
        if index:
            print(f"  {name}: {value} -> static table index {index}")
        else:
            print(f"  {name}: {value} -> literal (not in static table)")


if __name__ == "__main__":
    main()

File: phases/09-computer-networks/13-http-1-1-http-2-wire-format-and-multiplexing/code/main.rs
use std::collections::HashMap;
use std::io::{BufRead, BufReader, Read, Write};
use std::net::TcpListener;

// HTTP/1.1 Request structure
#[derive(Debug, Clone)]
pub struct HttpRequest {
    pub method: String,
    pub path: String,
    pub version: String,
    pub headers: HashMap<String, String>,
    pub body: Vec<u8>,
}

// HTTP/1.1 Response structure
#[derive(Debug, Clone)]
pub struct HttpResponse {
    pub status: u16,
    pub reason: String,
    pub headers: HashMap<String, String>,
    pub body: Vec<u8>,
}

/// Parse raw bytes into an HTTP/1.1 request.
/// Format: METHOD /path HTTP/1.1\r\nHeader: Value\r\n...\r\n\r\n<body>
pub fn parse_http_request(data: &[u8]) -> Result<HttpRequest, String> {
    let mut reader = BufReader::new(data);
    let mut request_line = String::new();
    reader
        .read_line(&mut request_line)
        .map_err(|e| format!("Failed to read request line: {e}"))?;

    let parts: Vec<&str> = request_line.trim().split_whitespace().collect();
    if parts.len() != 3 {
        return Err("Invalid request line".to_string());
    }

    let method = parts[0].to_string();
    let path = parts[1].to_string();
    let version = parts[2].to_string();

    let mut headers = HashMap::new();
    loop {
        let mut line = String::new();
        reader
            .read_line(&mut line)
            .map_err(|e| format!("Failed to read header: {e}"))?;
        let line = line.trim();
        if line.is_empty() {
            break;
        }
        if let Some((key, value)) = line.split_once(':') {
            headers.insert(key.trim().to_lowercase(), value.trim().to_string());
        }
    }

    let mut body = Vec::new();
    if let Some(content_length) = headers.get("content-length") {
        if let Ok(len) = content_length.parse::<usize>() {
            body.resize(len, 0);
            reader
                .read_exact(&mut body)
                .map_err(|e| format!("Failed to read body: {e}"))?;
        }
    }

    Ok(HttpRequest {
        method,
        path,
        version,
        headers,
        body,
    })
}

/// Serialize an HTTP response to bytes.
pub fn build_http_response(response: &HttpResponse) -> Vec<u8> {
    let mut buf = Vec::new();
    write!(
        &mut buf,
        "HTTP/1.1 {} {}\r\n",
        response.status, response.reason
    )
    .unwrap();

    for (key, value) in &response.headers {
        write!(&mut buf, "{key}: {value}\r\n").unwrap();
    }
    buf.extend_from_slice(b"\r\n");
    buf.extend_from_slice(&response.body);
    buf
}

// HTTP/2 frame types
pub const H2_FRAME_DATA: u8 = 0x0;
pub const H2_FRAME_HEADERS: u8 = 0x1;
pub const H2_FRAME_SETTINGS: u8 = 0x4;
pub const H2_FRAME_PING: u8 = 0x6;
pub const H2_FRAME_GOAWAY: u8 = 0x7;

/// Parsed HTTP/2 frame
#[derive(Debug, Clone)]
pub struct H2Frame {
    pub length: u32,
    pub frame_type: u8,
    pub flags: u8,
    pub stream_id: u32,
    pub payload: Vec<u8>,
}

impl H2Frame {
    pub fn frame_type_name(&self) -> &str {
        match self.frame_type {
            H2_FRAME_DATA => "DATA",
            H2_FRAME_HEADERS => "HEADERS",
            0x2 => "PRIORITY",
            0x3 => "RST_STREAM",
            H2_FRAME_SETTINGS => "SETTINGS",
            0x5 => "PUSH_PROMISE",
            H2_FRAME_PING => "PING",
            H2_FRAME_GOAWAY => "GOAWAY",
            0x8 => "WINDOW_UPDATE",
            0x9 => "CONTINUATION",
            _ => "UNKNOWN",
        }
    }
}

/// HTTP/2 connection preface (magic string sent by client)
pub const H2_PREFACE: &[u8] = b"PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n";

/// Parse an HTTP/2 frame from raw bytes.
/// Frame header: 3 bytes length + 1 byte type + 1 byte flags + 4 bytes stream ID (1 bit reserved)
pub fn parse_h2_frame(data: &[u8]) -> Result<(H2Frame, usize), String> {
    if data.len() < 9 {
        return Err("Not enough data for frame header".to_string());
    }

    let length = u32::from_be_bytes([0, data[0], data[1], data[2]]);
    let frame_type = data[3];
    let flags = data[4];
    let stream_id = u32::from_be_bytes([data[5] & 0x7F, data[6], data[7], data[8]]);

    let total = 9 + length as usize;
    if data.len() < total {
        return Err(format!(
            "Need {total} bytes but only have {}",
            data.len()
        ));
    }

    let payload = data[9..total].to_vec();

    Ok((
        H2Frame {
            length,
            frame_type,
            flags,
            stream_id,
            payload,
        },
        total,
    ))
}

/// Build an HTTP/2 frame header + payload.
pub fn build_h2_frame(frame_type: u8, flags: u8, stream_id: u32, payload: &[u8]) -> Vec<u8> {
    let length = payload.len() as u32;
    let mut buf = Vec::with_capacity(9 + payload.len());

    buf.push(((length >> 16) & 0xFF) as u8);
    buf.push(((length >> 8) & 0xFF) as u8);
    buf.push((length & 0xFF) as u8);
    buf.push(frame_type);
    buf.push(flags);
    buf.extend_from_slice(&(stream_id & 0x7FFF_FFFF).to_be_bytes());
    buf.extend_from_slice(payload);
    buf
}

/// Build a SETTINGS frame with default values.
pub fn build_settings_frame() -> Vec<u8> {
    // SETTINGS: id(2) + value(4) for each setting
    let mut payload = Vec::new();
    // SETTINGS_MAX_CONCURRENT_STREAMS = 3
    payload.extend_from_slice(&0x0003u16.to_be_bytes());
    payload.extend_from_slice(&100u32.to_be_bytes());
    // SETTINGS_INITIAL_WINDOW_SIZE = 4
    payload.extend_from_slice(&0x0004u16.to_be_bytes());
    payload.extend_from_slice(&65535u32.to_be_bytes());
    // SETTINGS_MAX_FRAME_SIZE = 5
    payload.extend_from_slice(&0x0005u16.to_be_bytes());
    payload.extend_from_slice(&16384u32.to_be_bytes());

    build_h2_frame(H2_FRAME_SETTINGS, 0x00, 0, &payload)
}

/// Build a PING frame.
pub fn build_ping_frame(opaque_data: [u8; 8]) -> Vec<u8> {
    build_h2_frame(H2_FRAME_PING, 0x00, 0, &opaque_data)
}

/// Build a PONG (PING ACK) response.
pub fn build_pong_frame(opaque_data: [u8; 8]) -> Vec<u8> {
    build_h2_frame(H2_FRAME_PING, 0x01, 0, &opaque_data)
}

fn main() {
    println!("=== HTTP/1.1 and HTTP/2 Protocol Parser ===\n");

    // Demo 1: Parse an HTTP/1.1 request
    let raw_request = b"GET /api/users HTTP/1.1\r\nHost: localhost:8080\r\nAccept: application/json\r\nConnection: keep-alive\r\n\r\n";
    match parse_http_request(raw_request) {
        Ok(req) => {
            println!("Parsed HTTP/1.1 Request:");
            println!("  Method:  {}", req.method);
            println!("  Path:    {}", req.path);
            println!("  Version: {}", req.version);
            println!("  Headers:");
            for (k, v) in &req.headers {
                println!("    {k}: {v}");
            }
        }
        Err(e) => eprintln!("Parse error: {e}"),
    }

    // Demo 2: Build an HTTP/1.1 response
    let mut resp_headers = HashMap::new();
    resp_headers.insert("Content-Type".into(), "application/json".into());
    resp_headers.insert("Content-Length".into(), "27".into());
    let response = HttpResponse {
        status: 200,
        reason: "OK".into(),
        headers: resp_headers,
        body: b"{\"users\":[\"alice\",\"bob\"]}".to_vec(),
    };
    let resp_bytes = build_http_response(&response);
    println!("\nBuilt HTTP/1.1 Response:");
    println!("{}", String::from_utf8_lossy(&resp_bytes));

    // Demo 3: Parse HTTP/2 frames
    println!("HTTP/2 Frame Examples:");

    // Connection preface
    println!(
        "\n  Connection preface: {:?}",
        String::from_utf8_lossy(H2_PREFACE)
    );

    // SETTINGS frame
    let settings = build_settings_frame();
    match parse_h2_frame(&settings) {
        Ok((frame, _)) => {
            println!("\n  Parsed SETTINGS frame:");
            println!("    Length:     {}", frame.length);
            println!("    Type:      {} ({})", frame.frame_type, frame.frame_type_name());
            println!("    Flags:     0x{:02X}", frame.flags);
            println!("    Stream ID: {}", frame.stream_id);
        }
        Err(e) => eprintln!("  Frame parse error: {e}"),
    }

    // PING frame
    let ping_data: [u8; 8] = [0xDE, 0xAD, 0xBE, 0xEF, 0x01, 0x02, 0x03, 0x04];
    let ping = build_ping_frame(ping_data);
    match parse_h2_frame(&ping) {
        Ok((frame, _)) => {
            println!("\n  Parsed PING frame:");
            println!("    Length:     {}", frame.length);
            println!("    Type:      {} ({})", frame.frame_type, frame.frame_type_name());
            println!("    Stream ID: {}", frame.stream_id);
            println!("    Payload:   {:02X?}", frame.payload);
        }
        Err(e) => eprintln!("  Frame parse error: {e}"),
    }

    // PONG response
    let pong = build_pong_frame(ping_data);
    match parse_h2_frame(&pong) {
        Ok((frame, _)) => {
            println!("\n  Parsed PING ACK frame:");
            println!("    Flags:     0x{:02X} (ACK={})", frame.flags, frame.flags & 0x01);
        }
        Err(e) => eprintln!("  Frame parse error: {e}"),
    }
}
